Sources/CoreMetrics/Metrics.swift

```diff
@@ -12,220 +12,230 @@
 //
 //===----------------------------------------------------------------------===//
 
-public protocol Counter: AnyObject {
+public protocol CounterHandler: AnyObject {
     func increment<DataType: BinaryInteger>(_ value: DataType)
 }
 
-public extension Counter {
+public class Counter: CounterHandler {
+    @usableFromInline
+    var handler: CounterHandler
+    public let label: String
+    public let dimensions: [(String, String)]
+
+    // this method is public to provide an escape hatch for situations one must use a custom factory instead of the gloabl one
+    // we do not expect this API to be used in normal circumstances, so if you find yourself using it make sure its for a good reason
+    public init(label: String, dimensions: [(String, String)], handler: CounterHandler) {
+        self.label = label
+        self.dimensions = dimensions
+        self.handler = handler
+    }
+
     @inlinable
-    func increment() {
+    public func increment<DataType: BinaryInteger>(_ value: DataType) {
+        self.handler.increment(value)
+    }
+
+    @inlinable
+    public func increment() {
         self.increment(1)
     }
 }
 
-public protocol Recorder: AnyObject {
+public extension Counter {
+    public convenience init(label: String, dimensions: [(String, String)] = []) {
+        let handler = MetricsSystem.handler.makeCounter(label: label, dimensions: dimensions)
+        self.init(label: label, dimensions: dimensions, handler: handler)
+    }
+}
+
+public extension Counter {
+    @inlinable
+    public static func `do`(label: String, dimensions: [(String, String)] = [], body: (Counter) -> Void) {
+        body(Counter(label: label, dimensions: dimensions))
```

Comment:
Just note that here we'd do the destroy immediately when the lifecycle hooks land I guess...?

Reply:
yes i’d like to deal with lifecycle after these API changes land